pkg/operator/controller/externaldns/deployment.go

```diff
@@ -0,0 +1,224 @@
+		}
+		depl.Spec.Template.Spec.Containers = append(depl.Spec.Template.Spec.Containers, *container)
+	}
+
+	return depl, nil
+}
+
+// WIP function for generating container specs for one container at a time.
+func populateExternalDNSContainer(image string, metricsPort int, zone string, externalDNS *operatorv1alpha1.ExternalDNS) (*corev1.Container, error) {
+	name := "externaldns-" + zone
+
+	args := []string{
+		fmt.Sprintf("--metrics-address=127.0.0.1:%d", metricsPort),
+		fmt.Sprintf("--text-owner-id=externaldns-%s", externalDNS.Name),
+		fmt.Sprintf("--zone-id-filter=%s", zone),
+		"--policy=sync",
+		"--registry=txt",
+		"--log-level=debug",
+	}
+
+	provider, ok := providerStringTable[externalDNS.Spec.Provider.Type]
+	if !ok {
+		return nil, fmt.Errorf("%s is not a supported provider", externalDNS.Spec.Provider.Type)
+	}
+	args = append(args, fmt.Sprintf("--provider=%s", provider))
+
+	//TODO: Add provider credentials logic
+
+	source, ok := sourceStringTable[externalDNS.Spec.Source.Type]
+	if !ok {
+		return nil, fmt.Errorf("%s is not a supported source type", externalDNS.Spec.Source.Type)
+	}
+	args = append(args, fmt.Sprintf("--source=%s", source))
+
+	if externalDNS.Spec.Source.Namespace != nil && len(*externalDNS.Spec.Source.Namespace) > 0 {
+		args = append(args, fmt.Sprintf("--namespace=%s", *externalDNS.Spec.Source.Namespace))
+	}
+
+	if externalDNS.Spec.Source.Service != nil && len(externalDNS.Spec.Source.Service.ServiceType) > 0 {
+		serviceTypeFilter, publishInternal := "", false
+		for i, serviceType := range externalDNS.Spec.Source.Service.ServiceType {
+			serviceTypeFilter += string(serviceType)
+			// avoid having a trailing comma
+			if len(externalDNS.Spec.Source.Service.ServiceType)-1 > i {
+				serviceTypeFilter += ","
+			}
+			args = append(args, fmt.Sprintf("--service-type-filter=%s", serviceTypeFilter))
+
+			if serviceType == corev1.ServiceTypeClusterIP {
+				publishInternal = true
+			}
+		}
+		if publishInternal {
+			args = append(args, "--publish-internal-services")
+		}
+	}
+
+	if externalDNS.Spec.Source.HostnameAnnotationPolicy == operatorv1alpha1.HostnameAnnotationPolicyIgnore {
+		args = append(args, "--ignore-hostname-annotation")
+	}
+
+	//TODO: Add logic for the CRD source.
+
+	return &corev1.Container{
+		Name:                     name,
+		Image:                    image,
+		ImagePullPolicy:          corev1.PullIfNotPresent,
+		TerminationMessagePolicy: corev1.TerminationMessageFallbackToLogsOnError,
+		Args:                     args,
+	}, nil
+}
+
+// createExternalDNSDeployment creates the given deployment using the reconciler's client.
+func (r *reconciler) createExternalDNSDeployment(ctx context.Context, depl *appsv1.Deployment) error {
+	if err := r.client.Create(ctx, depl); err != nil {
+		return fmt.Errorf("failed to create externalDNS deployment %s/%s: %w", depl.Namespace, depl.Name, err)
+	}
+	r.log.Info("created externalDNS deployment", "namespace", depl.Namespace, "name", depl.Name)
```

Comment:
For Info logs, better to use klog and the flags for verbosity: `klog.V(4).Infof`

Reply:
`controller-runtime` based operators and controller use [logr](https://pkg.go.dev/github.com/go-logr/logr) compatible structured logging, not `klog`. About the verbosity level, here were are logging at the default level (without specifying `V()`) so that this will be seen in both [development or production](https://pkg.go.dev/sigs.k8s.io/controller-runtime@v0.9.5/pkg/log/zap#Options).